### PR TITLE
use more FDs, avoid race conditions on active fs

### PR DESCRIPTION
**Pull request by ThomasWaldmann**
guess this is mostly done, fixes issues mentioned in #908 and #1038.

### CODE REVIEW

File: src/borg/archive.py [hunk 1]
```diff
@@ -195,6 +194,22 @@ def backup_io_iter(iterator):
...
+        fd = os_open(path=path, parent_fd=parent_fd, name=name, flags=flags, noatime=noatime)
+    try:
+        yield fd
+    finally:
+        try:
+            os.close(fd)
+        except OSError as err:
+            if err.errno == errno.ENODEV and sys.platform == 'darwin':
+                # macOS raises this when we try to close a character device
+                pass
```
Comment (ThomasWaldmann) [inline src/borg/archive.py]: looks like macOS not only has issues closing chr devices, but also opening blk devices.
maybe we need other open flags?

**Comment by ThomasWaldmann** [comment]
this is how a borg create (py36) with the code of this PR currently looks like:
```
# 6 is the fd of the "docs" directory, now backing up "3rd_party" dir and contents:

newfstatat(6, "3rd_party", {st_mode=S_IFDIR|0775, st_size=4096, ...}, AT_SYMLINK_NOFOLLOW) = 0
openat(6, "3rd_party", O_RDONLY|O_NOFOLLOW|O_NOATIME|O_CLOEXEC|O_DIRECTORY) = 7
fstat(7, {st_mode=S_IFDIR|0775, st_size=4096, ...}) = 0
flistxattr(7, "", 4096)                 = 0
ioctl(7, FS_IOC_GETFLAGS, 0x7ffe824aff1c) = 0
fgetxattr(7, "system.posix_acl_access", NULL, 0) = -1 ENODATA (No data available)
fgetxattr(7, "system.posix_acl_default", NULL, 0) = -1 ENODATA (No data available)
openat(AT_FDCWD, "docs/3rd_party", O_RDONLY|O_NONBLOCK|O_CLOEXEC|O_DIRECTORY) = 9
fstat(9, {st_mode=S_IFDIR|0775, st_size=4096, ...}) = 0
getdents(9, /* 6 entries */, 32768)     = 160
getdents(9, /* 0 entries */, 32768)     = 0
close(9)                                = 0

newfstatat(7, "README", {st_mode=S_IFREG|0664, st_size=215, ...}, AT_SYMLINK_NOFOLLOW) = 0
openat(7, "README", O_RDONLY|O_NOCTTY|O_NONBLOCK|O_NOATIME|O_CLOEXEC) = 9
fstat(9, {st_mode=S_IFREG|0664, st_size=215, ...}) = 0
read(9, "Here we store 3rd party document"..., 8388608) = 215
fadvise64(9, 0, 0, POSIX_FADV_DONTNEED) = 0
read(9, "", 8388393)                    = 0
fadvise64(9, 0, 0, POSIX_FADV_DONTNEED) = 0
flistxattr(9, "", 4096)                 = 0
ioctl(9, FS_IOC_GETFLAGS, 0x7ffe824afc7c) = 0
fgetxattr(9, "system.posix_acl_access", NULL, 0) = -1 ENODATA (No data available)
fgetxattr(9, "system.posix_acl_default", NULL, 0) = -1 ENODATA (No data available)
close(9)                                = 0

# (... all the stuff in there ...)

close(7)                                = 0
```

**Comment by ThomasWaldmann** [comment]
same with py37:
```
newfstatat(6, "3rd_party", {st_mode=S_IFDIR|0775, st_size=4096, ...}, AT_SYMLINK_NOFOLLOW) = 0
openat(6, "3rd_party", O_RDONLY|O_NOFOLLOW|O_NOATIME|O_CLOEXEC|O_DIRECTORY) = 7
fstat(7, {st_mode=S_IFDIR|0775, st_size=4096, ...}) = 0
flistxattr(7, "", 4096)                 = 0
ioctl(7, FS_IOC_GETFLAGS, 0x7fff0a3e595c) = 0
fgetxattr(7, "system.posix_acl_access", NULL, 0) = -1 ENODATA (No data available)
fgetxattr(7, "system.posix_acl_default", NULL, 0) = -1 ENODATA (No data available)
fcntl(7, F_DUPFD_CLOEXEC, 0)            = 9
fstat(9, {st_mode=S_IFDIR|0775, st_size=4096, ...}) = 0
fcntl(9, F_GETFL)                       = 0x78000 (flags O_RDONLY|O_LARGEFILE|O_NOFOLLOW|O_NOATIME|O_DIRECTORY)
fcntl(9, F_SETFD, FD_CLOEXEC)           = 0
getdents(9, /* 6 entries */, 32768)     = 160
getdents(9, /* 0 entries */, 32768)     = 0
lseek(9, 0, SEEK_SET)                   = 0
close(9)                                = 0

newfstatat(7, "README", {st_mode=S_IFREG|0664, st_size=215, ...}, AT_SYMLINK_NOFOLLOW) = 0
openat(7, "README", O_RDONLY|O_NOCTTY|O_NONBLOCK|O_NOATIME|O_CLOEXEC) = 9
fstat(9, {st_mode=S_IFREG|0664, st_size=215, ...}) = 0
read(9, "Here we store 3rd party document"..., 8388608) = 215
fadvise64(9, 0, 0, POSIX_FADV_DONTNEED) = 0
read(9, "", 8388393)                    = 0
fadvise64(9, 0, 0, POSIX_FADV_DONTNEED) = 0
flistxattr(9, "", 4096)                 = 0
ioctl(9, FS_IOC_GETFLAGS, 0x7fff0a3e56dc) = 0
fgetxattr(9, "system.posix_acl_access", NULL, 0) = -1 ENODATA (No data available)
fgetxattr(9, "system.posix_acl_default", NULL, 0) = -1 ENODATA (No data available)
close(9)                                = 0
```

File: src/borg/archive.py (hunk 1 above)
Comment (ThomasWaldmann) [reply src/borg/archive.py]: solved by not opening devices any more.

**Comment by ThomasWaldmann** [comment]
@verygreen ah, right. guess one can not change the file type of an existing inode (except with a sector editor maybe? :-) ).

is there any way / chance to kill the old inode and get a new one with the same inode number?

if not, considering this code is called once per file (fs item), I'll remove the file type check again.

**Comment by ThomasWaldmann** [comment]
(we also need to consider fs with synthetic inode numbers, like network fs for that)

**Comment by verygreen** [comment]
yes, technically inode numbers could be resused, so if you want to be 100% sure this did not happen, you need to look at inode number and generation, but generation apparently is not exposed outside of the kernel for whatever reason.

That said conditions of inode reuse in relatively short time should still be very rare even on synthetic fses (esp. there since there they typically just do like a constantly increasing counter).

Here is a real world example of inode reuse: https://github.com/GoogleCloudPlatform/gcsfuse/issues/57

On ext3/4 the fs tries to avoid reusing recently deleted inodes unless you really don't have any other left. Probably many other filesystems do this too.

**Pull request merged**